File: data/derived_cot/rq3_prompting/results/translation/r1/py_to_cpp/code_output/BigNumCalculator.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cstddef>

class BigNumCalculator {
public:
    static std::string add(const std::string& num1, const std::string& num2) {
        size_t max_len = std::max(num1.length(), num2.length());
        std::string a = std::string(max_len - num1.length(), '0') + num1;
        std::string b = std::string(max_len - num2.length(), '0') + num2;

        int carry = 0;
        std::string result;
        for (int i = static_cast<int>(max_len) - 1; i >= 0; --i) {
            int digit_sum = (a[i] - '0') + (b[i] - '0') + carry;
            carry = digit_sum / 10;
            int digit = digit_sum % 10;
            result = char('0' + digit) + result;
        }
        if (carry > 0) {
            result = char('0' + carry) + result;
        }
        return result;
    }

    static std::string subtract(const std::string& num1, const std::string& num2) {
        std::string a = num1;
        std::string b = num2;
        bool negative = false;

        if (a.length() < b.length()) {
            std::swap(a, b);
            negative = true;
        } else if (a.length() == b.length() && a < b) {
            std::swap(a, b);
            negative = true;
        }

        size_t max_len = std::max(a.length(), b.length());
        a = std::string(max_len - a.length(), '0') + a;
        b = std::string(max_len - b.length(), '0') + b;

        int borrow = 0;
        std::string result;
        for (int i = static_cast<int>(max_len) - 1; i >= 0; --i) {
            int digit_diff = (a[i] - '0') - (b[i] - '0') - borrow;
            if (digit_diff < 0) {
                digit_diff += 10;
                borrow = 1;
            } else {
                borrow = 0;
            }
            result = char('0' + digit_diff) + result;
        }

        // Remove leading zeros
        while (result.length() > 1 && result[0] == '0') {
            result.erase(result.begin());
        }

        if (negative) {
            result = '-' + result;
        }
        return result;
    }

    static std::string multiply(const std::string& num1, const std::string& num2) {
        size_t len1 = num1.length();
        size_t len2 = num2.length();
        std::vector<int> result(len1 + len2, 0);

        for (int i = static_cast<int>(len1) - 1; i >= 0; --i) {
            for (int j = static_cast<int>(len2) - 1; j >= 0; --j) {
                int mul = (num1[i] - '0') * (num2[j] - '0');
                int p1 = i + j;
                int p2 = i + j + 1;
                int total = mul + result[p2];
                result[p1] += total / 10;
                result[p2] = total % 10;
            }
        }

        // Skip leading zeros
        size_t start = 0;
        while (start < result.size() - 1 && result[start] == 0) {
            ++start;
        }

        std::string out;
        for (size_t k = start; k < result.size(); ++k) {
            out += char('0' + result[k]);
        }
        return out;
    }
};

File: data/derived_cot/rq3_prompting/results/translation/r1/cpp_to_py/code_output/BigNumCalculator.py
class BigNumCalculator:
    @staticmethod
    def add(num1: str, num2: str) -> str:
        max_len = max(len(num1), len(num2))
        num1_padded = '0' * (max_len - len(num1)) + num1
        num2_padded = '0' * (max_len - len(num2)) + num2

        carry = 0
        result_chars = []
        for i in range(max_len - 1, -1, -1):
            digit_sum = int(num1_padded[i]) + int(num2_padded[i]) + carry
            carry = digit_sum // 10
            digit = digit_sum % 10
            result_chars.insert(0, chr(digit + ord('0')))

        if carry > 0:
            result_chars.insert(0, chr(carry + ord('0')))

        return ''.join(result_chars)

    @staticmethod
    def subtract(num1: str, num2: str) -> str:
        n1 = num1
        n2 = num2
        negative = False

        # Ensure n1 >= n2 (absolute value)
        if len(n1) < len(n2) or (len(n1) == len(n2) and n1 < n2):
            n1, n2 = n2, n1
            negative = True

        max_len = max(len(n1), len(n2))
        n1 = '0' * (max_len - len(n1)) + n1
        n2 = '0' * (max_len - len(n2)) + n2

        borrow = 0
        result_chars = []
        for i in range(max_len - 1, -1, -1):
            digit_diff = int(n1[i]) - int(n2[i]) - borrow
            if digit_diff < 0:
                digit_diff += 10
                borrow = 1
            else:
                borrow = 0
            result_chars.insert(0, chr(digit_diff + ord('0')))

        # Remove leading zeros (but keep at least one digit)
        while len(result_chars) > 1 and result_chars[0] == '0':
            result_chars.pop(0)

        if negative:
            result_chars.insert(0, '-')

        return ''.join(result_chars)

    @staticmethod
    def multiply(num1: str, num2: str) -> str:
        len1, len2 = len(num1), len(num2)
        result = [0] * (len1 + len2)

        for i in range(len1 - 1, -1, -1):
            for j in range(len2 - 1, -1, -1):
                mul = int(num1[i]) * int(num2[j])
                p1 = i + j
                p2 = i + j + 1
                total = mul + result[p2]
                result[p1] += total // 10
                result[p2] = total % 10

        # Build result string, skip leading zeros
        parts = []
        leading_zero = True
        for digit in result:
            if digit == 0 and leading_zero:
                continue
            leading_zero = False
            parts.append(str(digit))

        return ''.join(parts) if parts else '0'